src/SynthetixJoin.sol

```diff
@@ -0,0 +1,104 @@
+// SPDX-License-Identifier: AGPL-3.0-or-later
+// Copyright (C) 2021 Dai Foundation
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+pragma solidity 0.6.12;
+
+import "./CropJoin.sol";
+
+interface StakingRewardsLike {
+    function rewardsToken() external view returns (address);
+    function stakingToken() external view returns (address);
+    function stake(uint256) external;
+    function withdraw(uint256) external;
+    function getReward() external;
+}
+
+// Join adapter for the Synthetix Staking Rewards contract (used by Uniswap V2, LIDO, etc)
+contract SynthetixJoinImp is CropJoinImp {
+
+    StakingRewardsLike immutable public pool;
+
+    /**
+        @param vat_                 MCD_VAT DSS core accounting module
+        @param ilk_                 Collateral type
+        @param gem_                 The collateral LP token address
+        @param bonus_               The rewards token contract address.
+        @param pool_                The staking rewards pool.
+    */
+    constructor(
+        address vat_,
+        bytes32 ilk_,
+        address gem_,
+        address bonus_,
+        address pool_
+    )
+        public
+        CropJoinImp(vat_, ilk_, gem_, bonus_)
+    {
+        // Sanity checks
+        require(StakingRewardsLike(pool_).rewardsToken() == bonus_, "SynthetixJoin/bonus-mismatch");
+        require(StakingRewardsLike(pool_).stakingToken() == gem_, "SynthetixJoin/gem-mismatch");
+
+        pool = StakingRewardsLike(pool_);
+    }
+
+    function init() external {
+        gem.approve(address(pool), type(uint256).max);
+    }
```

Comment:
Why not do the approval in the constructor? There's no access control here so I don't any argument based on wanting to separate contract deployment from farming commencement.

Reply:
Storage updates need to be done outside of the constructor because the storage is in the proxy contract and not the current implementation.

Reply:
Ah okay didn't realize this would be behind a proxy.